### PR TITLE
refactor: prepare for stable metrics static analysis by defining metric labels with same format

### LINKED ISSUE #1833: Presubmit checks for stable metrics
**What would you like to be added**:
A presubmit to validate stable metrics don't change metric name, add labels or delete labels. 

Similar to [k8s stable metrics presubmit](https://github.com/kubernetes/kubernetes/tree/master/test/instrumentation).

**Why is this needed**:
Guarantee stable metrics are stable.

Due to lack of presubmit check, some stable metrics are changed. For example, 
- stable metric `kube_persistentvolume_info` is changed in https://github.com/kubernetes/kube-state-metrics/pull/1814
- stable metric `kube_service_status_load_balancer_ingress` is changed in https://github.com/kubernetes/kube-state-metrics/pull/1687

**Describe the solution you'd like**
1. Import k8s metrics stability framework https://github.com/kubernetes/kube-state-metrics/pull/1844 and https://github.com/kubernetes/kube-state-metrics/pull/1976
2. Presubmit check https://github.com/kubernetes/kube-state-metrics/pull/1836

**Additional context**

### PR DESCRIPTION
**Pull request by CatherineF-dev**
…ls with same format

**What this PR does / why we need it**: 

KSM metrics are defined freely, which increases the difficulties of static analysis.

Will add a static analysis to make sure each metric defines label in this way:

1. SetLabelKeys appears once
2. The values is []string{*}
```
labelKeys := []string{"container"}

metric.SetLabelKeys(ms, labelKeys)
```

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)* N/A

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1833

[x] certificatesigningrequest.go
[x] clusterrolebinding.go
[x] clusterrole.go
[x] configmap.go
[x] cronjob.go
[x] daemonset.go
[x] deployment.go
[x] endpoint.go
[x] endpointslice.go
[x] horizontalpodautoscaler.go
[x] ingressclass.go
[x] ingress.go
[x] job.go
[x] lease.go
[x] limitrange.go
[x] mutatingwebhookconfiguration.go
[x] namespace.go
[x] networkpolicy.go
[x] node.go
[x] persistentvolumeclaim.go
[x] persistentvolume.go
[x] poddisruptionbudget.go
[x] pod.go
[x] replicaset.go
[x] replicationcontroller.go
[x] resourcequota.go
[x] rolebinding.go
[x] role.go
[x] secret.go
[x] serviceaccount.go
[x] service.go
[x] statefulset.go
[x] storageclass.go
[x] validatingwebhookconfiguration.go
[x] volumeattachment.go

### CODE REVIEW

**Comment by CatherineF-dev** [comment]
cc @dgrisonnet to review

**Comment by dgrisonnet** [comment]
Forcing a particular coding style will make it a bit harder for contributors, but if we really have to do that, I am fine with it.
Could you perhaps explain what is your current approach that requires this change so that I can better understand your reasoning?

Ideally if we go that route, we should reflect it in https://github.com/kubernetes/kube-state-metrics/blob/main/docs/developer/guide.md, to documented the new coding style.

Also, there are a bunch more areas of the code that has this kind of definition, so it would be better to do it in one swoop: https://github.com/search?q=repo%3Akubernetes%2Fkube-state-metrics+LabelKeys%3A+%5B%5Dstring&type=code

**Comment by CatherineF-dev** [comment]
> Could you perhaps explain what is your current approach that requires this change so that I can better understand your reasoning?

Similar to static analysis in k/k repo, it will extracts all metric labels for each metric and generate a metric file containing metric name and metric labels.  During presubmit, it regenerates this file and compare with previously generated file.

Since the way of defining metric labels are constrained a little, it's easier to extract metric labels.

For example, it will get label "container" for this metric.
```
			for _, metric := range ms {
				metric.LabelKeys = []string{"container"}
			}
			return &metric.Family{
				Metrics: ms,
			}
```

**Comment by CatherineF-dev** [comment]
> Also, there are a bunch more areas of the code that has this kind of definition, so it would be better to do it in one swoop: https://github.com/search?q=repo%3Akubernetes%2Fkube-state-metrics+LabelKeys%3A+%5B%5Dstring&type=code

Okay, will update other files:

```
configmap.go
node.go
clusterrole.go
ingress.go
clusterrolebinding.go
poddisruptionbudget.go
ingressclass.go
persistentvolume.go
serviceaccount.go
endpoint.go
pod.go
storageclass.go
job.go
lease.go
daemonset.go
role.go
builder.go
resourcequota.go
replicationcontroller.go
volumeattachment.go
service.go
secret.go
persistentvolumeclaim.go
replicaset.go
rolebinding.go
deployment.go
namespace.go
horizontalpodautoscaler.go
endpointslice.go
limitrange.go
statefulset.go
certificatesigningrequest.go
mutatingwebhookconfiguration.go
cronjob.go
validatingwebhookconfiguration.go
networkpolicy.go

```

**Comment by CatherineF-dev** [comment]
cc @mrueg for review.

**Comment by dgrisonnet** [comment]
> Similar to static analysis in k/k repo, it extracts all metric labels for each STABLE metric and generate a metric file containing metric name and metric labels. During presubmit, it regenerates this file and compare with previously generated file.

Yeah I know this is the general goal, but could you explain how this refactor makes it easier to parse?

**Comment by CatherineF-dev** [comment]
Before `a`: 
https://github.com/kubernetes/kube-state-metrics/blob/14e935a49fba50f81207b5fbec21e02da529c2cb/internal/store/pod.go#L136-L164


Now `b`:
https://github.com/kubernetes/kube-state-metrics/blob/5ff07b5cdb8286772f90b7df126dcebd789da28c/internal/store/pod.go#L137-L167

- It reduces depth from root node to metric label node.  For example, example `a` needs following this path to find labels `createPodContainerInfoFamilyGenerator > return > wrapPodFunc > for > for > ms > LabelKeys`. Example `b` only needs to follow this path `createPodContainerInfoFamilyGenerator > return > wrapPodFunc  > LabelKeys`

- It defines all non-shared labels in the beginning. In example `a`, more labelKeys can be added in the same function later.  

- Also improve readability that readers can find metric labels easily.

cc @dgrisonnet

**Comment by mrueg** [comment]
Similar thoughts as @dgrisonnet 
Additionally: Is this coding style enforceable via some conformance test / CI ?
I would also like to see an example static analysis tool in parallel to this effort.

File: internal/store/utils.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ import (
 	"k8s.io/kube-state-metrics/v2/pkg/options"
 )
 
+// SharedLabelKeys is the type for labelKeys which are shared in all metrics under one resource type
+type SharedLabelKeys []string
```
Comment (mrueg) [inline internal/store/utils.go:34]: What is the benefit of making this a type?

Comment (logicalhan) [reply internal/store/utils.go:34]: It's easier to parse using static analysis?

Comment (CatherineF-dev) [reply internal/store/utils.go:34]: Yes, it makes parsing easier.

Find variable with type SharedLabelKeys, then append these shared labels into each ingress metric.

https://github.com/kubernetes/kube-state-metrics/blob/c8c94159b10ca8c4ab55a0fb2dd0cfd96fd15ef6/internal/store/ingress.go#L41

**Comment by CatherineF-dev** [comment]
> Additionally: Is this coding style enforceable via some conformance test / CI ?

Yes, will add it later.

> I would also like to see an example static analysis tool in parallel to this effort.

Okay, in adding it.

**Comment by CatherineF-dev** [comment]
Will use https://github.com/kubernetes/kube-state-metrics/pull/2294 instead. It's an end-to-end example on verifying stable metrics.

**Pull request closed without merging**